sentry/src/main/java/io/sentry/clientreport/AtomicClientReportStorage.java

```diff
@@ -0,0 +1,80 @@
+package io.sentry.clientreport;
+
+import io.sentry.SentryLevel;
+import io.sentry.SentryOptions;
+import io.sentry.transport.DataCategory;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.atomic.AtomicLong;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+final class AtomicClientReportStorage implements ClientReportStorage {
+
+  private final @NotNull Map<ClientReportKey, AtomicLong> lostEventCounts;
+
+  public AtomicClientReportStorage() {
+    final Map<ClientReportKey, AtomicLong> modifyableEventCountsForInit = new ConcurrentHashMap<>();
+
+    for (final DiscardReason discardReason : DiscardReason.values()) {
+      for (final DataCategory category : DataCategory.values()) {
+        modifyableEventCountsForInit.put(
+            new ClientReportKey(discardReason.getReason(), category.getCategory()),
+            new AtomicLong(0));
+      }
+    }
+
+    lostEventCounts = Collections.unmodifiableMap(modifyableEventCountsForInit);
+  }
+
+  @Override
+  public void addCount(ClientReportKey key, Long count) {
+    final @Nullable AtomicLong quantity = lostEventCounts.get(key);
+
+    if (quantity != null) {
+      quantity.addAndGet(count);
+    }
+  }
+
+  @Override
+  public List<DiscardedEvent> resetCountsAndGet() {
+    final List<DiscardedEvent> discardedEvents = new ArrayList<>(lostEventCounts.size());
```

Comment:
Nice optimization of already setting the size of the ArrayList.

Reply:
Hmm not sure this still makes sense since the atomic storage has all possible combinations so the list will (almost) always be oversized as only items with quantity > 0 will make it to the list.

Reply:
I think it depends on the implementation of ArrayList. I thought that it comes with a default of 32 or something, but I was surprised that on Java 11 the default size is 0, and the array keeps growing by one element when you add a new element. I think I mixed the default size up with the HashMap.
So in our case, I think it would be better to not set the initial size to `lostEventCounts.size()` because of what you just mentioned. Anyways, the list is so small that I think it won't matter anyways.